Implement some additional set operations

bobeff opened the pull request:
* Conversions from `HashSet` to `OrderedSet` and vice versa.

* Inclusions and exclusions of members of any set type (`HashSet` or `OrderedSet`) to any other set type.

kaushalmodi commented on lib/pure/collections/sets.nim:
```diff
@@ -80,6 +80,8 @@ type
     ## <#initOrderedSet,int>`_ before calling other procs on it.
     data: OrderedKeyValuePairSeq[A]
     counter, first, last: int
+  AnySet1[T] = HashSet[T] | OrderedSet[T]
```
- Is it needed to have both AnySet1 and AnySet2? Why not use the same type for both args?
- Also, the convention is to use SomeFoo. E.g. we have SomeNumber, SomeInteger. So may be use SomeSet here?

bobeff replied:
* It is needed because the sets can be different, but not mandatory different. I tried `proc incl*[T](s: var AnySet[T], other: distinct AnySet[T])`, but unfortunately for some reason this requires mandatory different sets. I'm not sure whether this is by specification or it is a bug in the compiler.

* I will change it to `SomeSet1` and `SomeSet2`.

kaushalmodi replied:
Thanks. May be someone else can comment if it should be possible to use to the same type for the `incl`.

kaushalmodi commented:
Can you also update the changelog.md in this repo root for the changes, additions in this PR?

bobeff commented:
@kaushalmodi  Done.

kaushalmodi commented on lib/pure/collections/sets.nim:
```diff
@@ -730,17 +730,19 @@ proc incl*[A](s: var OrderedSet[A], key: A) =
     var
       values = toHashSet([1, 2, 3])
-      others = toOrderedSet([3, 4, 5])
+      others = toHashSet([3, 4, 5])
```
If SomeSet1/2 can be either of the 2 set types, why does this test need to change?

bobeff replied:
This is some artifact caused by trying different implementations. Before I wrote the generic on the set type procedure first I stated to write four different procedures. I will revert it.

bobeff replied:
It's fixed.

Araq commented on lib/pure/collections/sets.nim:
```diff
@@ -80,6 +80,8 @@ type
     data: OrderedKeyValuePairSeq[A]
     counter, first, last: int
+  SomeSet1[T] = HashSet[T] | OrderedSet[T]
+  SomeSet2[T] = HashSet[T] | OrderedSet[T]
```
The second argument should be ``Iterable[T]`` though.

bobeff replied:
But unfortunately still there is no such concept into the standard library. I can define it, but is it allowed to use concepts which are still an experimental feature?

Araq replied:
Yes, if it's tested well. They are experimental but in the worst case they will be replaced by something that can express the same.

bobeff replied:
Done. I also tried `proc incl*[T](s: SomeSet[T], others: varargs[Iterable[T]])` to allow usage like `someSet.incl(someIterable, someOtherIterable)`, but unfortunately it seems that currently `varargs` does not work with concept arguments.

Araq replied:
Well ``varargs[Iterable[T]]`` would be too much for my taste anyway.

bobeff commented:
I also added overload of `[]` operator for constant types.

dom96 commented on lib/pure/collections/sets.nim:
```diff
@@ -80,6 +80,10 @@ type
     data: OrderedKeyValuePairSeq[A]
     counter, first, last: int
+  SomeSet[T] = HashSet[T] | OrderedSet[T]
+  Iterable[T] = concept x
+    for value in x:
+      type(value) is T
```
Don't use concepts that aren't exported, how can the user of this API use this concept or at least see its definition?

Exporting this concept would require a separate PR (it should be added in a different module and used throughout the stdlib), so for this one you simply shouldn't use it IMO.

timotheecour replied:
without concepts (or enableif or anything equally capable), this is a recipie for introducing bloat like `proc toHashSet*[A](s: OrderedSet[A]): HashSet[A] =` (see my remark in https://github.com/nim-lang/Nim/pull/14481/files#r439715608), however I agree that the iterable concept should be in another PR; but this PR can't fly without these

dom96 reviewed (changes requested):
Nice additions, but `Iterable` should not be introduced.

Araq commented:
@dom96 I asked for the inclusion of `Iterable` because it's time we use some concepts. (We can always fix the syntax later ;-) )

timotheecour commented on lib/pure/collections/sets.nim:
```diff
@@ -227,6 +224,23 @@ proc toHashSet*[A](keys: openArray[A]): HashSet[A] =
   result = initHashSet[A](rightSize(keys.len))
   for key in items(keys): result.incl(key)
 
+proc toHashSet*[A](s: OrderedSet[A]): HashSet[A] =
```
this should be
```nim
proc toHashSet*[A](s: IterableLen[A]): HashSet[A]
```
and merged with `proc toHashSet*[A](keys: openArray[A]): HashSet[A] =`
because all you're using is that it's iterable and has a `len`
=> needs a new concept `IterableLen` which is generally useful here and in sequtils among others

bobeff replied:
Fixed. I also started a concepts module with the first two concepts used in the `sets` module: `Iterable[T]` and `IterableLen[T]`.

dom96 commented:
Concepts aren't ever strictly necessary, they just make it more clear what sort of capabilities the type that is passed in should have. @timotheecour is already asking for more concepts... I honestly feel that these need to be introduced with purpose across the whole stdlib not in `sets` of all places. You can just use any generic type variable instead of `Iterable` here.

Araq commented:
1. Don't remove the ``items`` inside generic code. (Nim bug.)
2. Mergable once the CIs are green.

Please bring this over the finish line! :-)

bobeff commented:
Unfortunately the new implementation with concepts has a breaking change
```nim
import sets, sugar
dump toHashSet[string](["test"])
```
gives
> Error: cannot instantiate: 'toHashSet[string]'; got 1 typeof(s) but expected 2

Now it must be:
```nim
import sets, sugar
dump toHashSet(["test"])
```
which compiles successfully.

I don't know why this happens, but it causes the Nimble build to fail. I can easily fix it there but other people code could be also broken.

Any suggestions?

Araq commented:
>  I can easily fix it there but other people code could be also broken. Any suggestions?

It's not good enough. Probably we need a concept related bugfix for this, will investigate.

zah commented:
The issue here is that once you add a concept to the signature, the function gains an additional implicit generic parameter. When you do explicit instantiation, Nim still requires you to provide concrete types for all generic parameters (implicit or not). This is an old and well known issue. Usually, the suggestion is to use typedesc parameters instead (they don't suffer from this issue), but that's not option here, because it's even more breaking change.

bobeff commented:
@Araq @zah Maybe implementation of partial generic parameters match and automatic deduction of the rest, like in C++ will be appropriate solution?

bobeff committed:
Implemet additional set operations

- Standard library concepts module is started with the `Iterable` and
  `IterableLen` concepts.

- All operations on `HashSet` and `OrderedSet` are implemented also for
  the other set type via generic interface working with an union type
  `SomeSet` and the newly added concepts `Iterable` and `IterableLen`.

bobeff commented:
Currently this is blocked by nim-lang/nim#14729.

timotheecour commented on lib/pure/concepts.nim:
```diff
@@ -0,0 +1,35 @@
+## across Nim's standard library.
+
+type
+  Iterable*[T] = concept x
```
we now have `iterable[T]` which allows matching called iterators, unlike this `Iterable` concept, eg:
```nim
iterator iota(n: int): int =
  for i in 0..<n: yield i
proc sum[T](a: iterable[T]): T = # would not work with `Iterable`
  for ai in a: result += ai
assert 3.iota.sum == 0+1+2
```
